File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/units/Units.java
package ch.rmy.android.statusbar_tacho.units;

public class Units {

    public static final Unit[] UNITS = {
            new KilometersPerHourUnit(),
            new MilesPerHourUnit(),
            new MetersPerSecondUnit(),
            new FeetPerSecondUnit(),
    };

    public static Unit deserializeUnit(int unitId) {
        return UNITS[unitId];
    }

    public static int serializeUnit(Unit unit) {
        for (int i = 0; i < UNITS.length; i++) {
            if (UNITS[i].getClass() == unit.getClass()) {
                return i;
            }
        }
        throw new IllegalArgumentException("Unknown unit");
    }

}

File: StatusBarSpeedometer/app/src/main/java/ch/rmy/android/statusbar_tacho/units/Units.kt
package ch.rmy.android.statusbar_tacho.units

object Units {

    val UNITS = arrayOf(KilometersPerHourUnit(), MilesPerHourUnit(), MetersPerSecondUnit(), FeetPerSecondUnit())

    fun deserializeUnit(unitId: Int): Unit {
        return UNITS[unitId]
    }

    fun serializeUnit(unit: Unit): Int {
        for (i in UNITS.indices) {
            if (UNITS[i].javaClass == unit.javaClass) {
                return i
            }
        }
        throw IllegalArgumentException("Unknown unit")
    }

}
